core/predicate_check_test.go

```diff
@@ -14,150 +13,255 @@ import (
 		},
-		"predicate with valid access list and non-empty PredicateGas passes": {
-			address: common.HexToAddress("0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"),
-			gas:     153000,
-			predicater: &mockPredicater{
-				predicateFunc: func(_ *precompileconfig.PredicateContext, b []byte) error {
-					if !bytes.Equal(b, common.Hash{1}.Bytes()) {
-						return fmt.Errorf("unexpected bytes: 0x%x", b)
-					}
-					return nil
+		"predicate returns gas err": {
+			gas: 53000,
+			createPredicates: func(t testing.TB) map[common.Address]precompileconfig.Predicater {
+				predicate := precompileconfig.NewMockPredicater(gomock.NewController(t))
+				arg := common.Hash{1}
+				predicate.EXPECT().PredicateGas(arg[:]).Return(uint64(0), testErr)
+				return map[common.Address]precompileconfig.Predicater{
+					addr1: predicate,
+				}
+			},
+			accessList: types.AccessList([]types.AccessTuple{
+				{
+					Address: addr1,
+					StorageKeys: []common.Hash{
+						{1},
+					},
 				},
-				predicateGasFunc: func(b []byte) (uint64, error) {
-					return 100_000, nil
+			}),
+			expectedErr: testErr,
+		},
+		"two predicates one named by access list returns non-empty": {
+			gas: 53000,
+			createPredicates: func(t testing.TB) map[common.Address]precompileconfig.Predicater {
+				predicate := precompileconfig.NewMockPredicater(gomock.NewController(t))
+				arg := common.Hash{1}
+				predicate.EXPECT().PredicateGas(arg[:]).Return(uint64(0), nil).Times(2)
+				predicate.EXPECT().VerifyPredicate(gomock.Any(), [][]byte{arg[:]}).Return(predicateResultBytes)
+				return map[common.Address]precompileconfig.Predicater{
+					addr1: predicate,
+					addr2: predicate,
+				}
+			},
+			accessList: types.AccessList([]types.AccessTuple{
+				{
+					Address: addr1,
+					StorageKeys: []common.Hash{
+						{1},
+					},
 				},
+			}),
+			expectedRes: map[common.Address][]byte{
+				addr1: predicateResultBytes,
+			},
+			expectedErr: nil,
+		},
+		"two predicates both named by access list returns non-empty": {
+			gas: 53000,
+			createPredicates: func(t testing.TB) map[common.Address]precompileconfig.Predicater {
+				predicate := precompileconfig.NewMockPredicater(gomock.NewController(t))
+				arg := common.Hash{1}
+				predicate.EXPECT().PredicateGas(arg[:]).Return(uint64(0), nil).Times(4)
+				predicate.EXPECT().VerifyPredicate(gomock.Any(), [][]byte{arg[:]}).Return(predicateResultBytes).Times(2)
+				return map[common.Address]precompileconfig.Predicater{
+					addr1: predicate,
+					addr2: predicate,
+				}
 			},
 			accessList: types.AccessList([]types.AccessTuple{
 				{
-					Address: common.HexToAddress("0x8db97C7cEcE249c2b98bDC0226Cc4C2A57BF52FC"),
+					Address: addr1,
+					StorageKeys: []common.Hash{
+						{1},
+					},
+				},
+				{
+					Address: addr2,
 					StorageKeys: []common.Hash{
 						{1},
 					},
 				},
 			}),
+			expectedRes: map[common.Address][]byte{
+				addr1: predicateResultBytes,
+				addr2: predicateResultBytes,
```

Comment:
nit: may be nice to verify each predicate returns its corresponding result instead of returning both the same.

Reply:
Updating